src-tauri/src/lib.rs

```diff
@@ -1,14 +1,193 @@
+    let client = reqwest::Client::new();
+
+    // Send a GET request to the GitHub API to get tag information
+    let response = match client.get(url).header("User-Agent", "reqwest").send().await {
+        Ok(resp) => resp,
+        Err(_) => return LatestVersions::new(true), // If request fails, return LatestVersions with error flag set
+    };
+
+    // Parse the response as JSON and convert it into a vector of TagResponse
+    let tags = match response.json::<Vec<TagResponse>>().await {
+        Ok(json) => json,
+        Err(_) => return LatestVersions::new(true), // If parsing fails, return LatestVersions with error flag set
+    };
+
+    let mut latest_versions = LatestVersions::new(false);
+
+    // Iterate over the tags in reverse order
+    for item in tags.into_iter().rev() {
+        if let Some(tag) = item.r#ref.as_str().strip_prefix("refs/tags/") {
+            // Determine the type of version and update the latest version if necessary
+            if tag.contains("nightly") {
+                if latest_versions.nightly.is_none() || latest_versions.nightly.as_deref().unwrap() < tag {
+                    latest_versions.nightly = Some(tag.to_string());
+                }
+            } else if tag.contains("alpha") {
+                if latest_versions.alpha.is_none() || latest_versions.alpha.as_deref().unwrap() < tag {
+                    latest_versions.alpha = Some(tag.to_string());
+                }
+            } else if tag.contains("beta") {
+                if latest_versions.beta.is_none() || latest_versions.beta.as_deref().unwrap() < tag {
+                    latest_versions.beta = Some(tag.to_string());
+                }
+            } else if tag.contains("rc") {
+                if latest_versions.rc.is_none() || latest_versions.rc.as_deref().unwrap() < tag {
+                    latest_versions.rc = Some(tag.to_string());
+                }
+            } else {
+                if latest_versions.stable.is_none() || latest_versions.stable.as_deref().unwrap() < tag {
+                    latest_versions.stable = Some(tag.to_string());
+                }
+            }
+
+            // If all versions are found, break out of the loop
+            if latest_versions.nightly.is_some()
+                && latest_versions.alpha.is_some()
+                && latest_versions.stable.is_some()
+                && latest_versions.beta.is_some()
+                && latest_versions.rc.is_some()
+            {
+                break;
+            }
+        }
+    }
+
+    latest_versions
+}
+
+#[derive(Serialize)]
+struct DownloadResult {
+    error: bool,
+    message: Option<String>,
+}
+
+impl DownloadResult {
+    fn new(error: bool, message: Option<String>) -> Self {
+        Self { error, message }
+    }
+}
+
+#[tauri::command]
+// Asynchronous function to download a specific release based on the version
+async fn download_release(version: &str) -> Result<DownloadResult, String> {
```

Comment:
Consider define a struct `struct Error { message: String }` intead of String? It's hard to determine if the progress is failed. We need a shared field like `status` or `error` or `success` to flag a function's result. We can just downcast the field to check if the function returned an error.